File: Java/JDI/jdi-uitest-mobile/src/main/java/com/epam/jdi/uitests/mobile/appium/elements/complex/table/Rows.java
package com.epam.jdi.uitests.mobile.appium.elements.complex.table;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.commons.map.MapArray;
import com.epam.jdi.uitests.core.interfaces.common.IText;
import com.epam.jdi.uitests.mobile.appium.elements.complex.table.interfaces.ICell;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.Collection;
import java.util.List;

import static com.epam.commons.LinqUtils.select;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

/**
 * Created by 12345 on 26.10.2014.
 */
public class Rows extends TableLine {
    public Rows() {
        hasHeader = false;
        elementIndex = ElementIndexType.Nums;
        headersLocator = By.xpath(".//tr/td[1]");
        defaultTemplate = By.xpath(".//tr[%s]/td");
    }

    protected List<WebElement> getHeadersAction() {
        return table.getWebElement().findElements(headersLocator);
    }

    protected List<WebElement> getFirstLine() {
        return table.columns().getLineAction(1);
    }

    protected int getCount() {
        List<WebElement> elements = getFirstLine();
        return elements != null ? elements.size() : 0;
    }

    public MapArray<String, MapArray<String, ICell>> get() {
        return new MapArray<>(headers(), key -> key, this::getRow);
    }

    ///

    public List<String> getRowValue(String rowName) {
        try {
            return select(table.rows().getLineAction(rowName), WebElement::getText);
        } catch (Exception | Error ex) {
            throw throwRowsException(rowName, ex.getMessage());
        }
    }

    public final MapArray<String, String> getRowAsText(String rowName) {
        return getRow(rowName).toMapArray(IText::getText);
    }

    public MapArray<String, ICell> cellsToRow(Collection<ICell> cells) {
        return new MapArray<>(cells,
                cell -> headers()[cell.columnNum() - 1],
                cell -> cell);
    }

    public MapArray<String, ICell> getRow(int rowNum) {
        if (count() < 0 || table.rows().count() < rowNum || rowNum <= 0)
            throw exception("Can't Get Row '%s'. [num] > ColumnsCount(%s).", rowNum, count());
        try {
            List<WebElement> webRow = getLineAction(rowNum);
            return new MapArray<>(table.columns().count(),
                    key -> table.columns().headers()[key],
                    value -> table.cell(webRow.get(value), new Column(value + 1), new Row(rowNum)));
        } catch (Exception | Error ex) {
            throw throwRowsException(Integer.toString(rowNum) + "", ex.getMessage());
        }
    }

    public List<String> getRowValue(int rowNum) {
        if (count() < 0 || count() < rowNum || rowNum <= 0)
            throw exception("Can't Get Row '%s'. [num] > ColumnsCount(%s).", rowNum, count());
        try {
            return select(table.rows().getLineAction(rowNum), WebElement::getText);
        } catch (Exception | Error ex) {
            throw throwRowsException(Integer.toString(rowNum), ex.getMessage());
        }
    }

    public final MapArray<String, String> getRowAsText(int rowNum) {
        return getRow(rowNum).toMapArray(IText::getText);
    }

    public final MapArray<String, ICell> getRow(String rowName) {
        try {
            String[] headers = table.columns().headers();
            return new MapArray<>(table.columns().count(),
                    key -> table.columns().headers[key],
                    value -> table.cell(getLineAction(rowName).get(value), new Column(headers[value]), new Row(rowName)));
        } catch (Exception | Error ex) {
            throw throwRowsException(rowName, ex.getMessage());
        }
    }

    private RuntimeException throwRowsException(String colName, String ex) {
        return exception("Can't Get Row '%s'. Exception: %s", colName, ex);
    }
}

File: Python/JDI/web/selenium/elements/complex/table/rows.py
from JDI.core.interfaces.element_index_type import ElementIndexType
from JDI.web.selenium.driver.timer import Time
from JDI.web.selenium.elements.api_interact.find_element_by import By
from JDI.web.selenium.elements.complex.table.column import Column
from JDI.web.selenium.elements.complex.table.row import Row
from JDI.web.selenium.elements.complex.table.table_line import TableLine


class Rows(TableLine):
    def __init__(self):
        self.has_header = False
        self.element_index = ElementIndexType.nums
        self.by_headers_locator = By.xpath(".//tr/td[1]")
        self.by_default_template = By.xpath(".//tr[%s]/td")

    def get_headers_action(self):
        return self.table.get_web_element().find_elements(self.by_headers_locator[0], self.by_headers_locator[1])

    def get_row_value(self, row_name):
        try:
            return list(map(lambda el: el.text, self.get_line_action(row_name)))
        except Exception as ex:
            raise ex

    def get_first_line(self):
        return self.table.columns.get_line_action(1)

    def get_row(self, val):
        try:
            cols_count = self.table.columns.get_count()
            web_row = Time.wait_result_by_conditions(lambda: self.get_line_action(val),
                                                        lambda els: len(els) == cols_count)
            res = list()
            if isinstance(val, int):
                if val <= 0:
                    raise Exception("Table indexes starts from 1")
                if self.get_count() < 0 or self.get_count() < val or val <=0:
                    raise Exception("Can't Get Row '{0}'. [num] > ColumnsCount({1}).".format(val, self.get_count()))
                for i in range(cols_count):
                    res.append(tuple([self.table.columns.get_headers()[i],self.table.cell(web_row[i], Column(i+1), Row(val))]))
            elif isinstance(val, str):
                    headers = self.table.columns.get_headers()
                    res = list()
                    for i in range(cols_count):
                        res.append(tuple([self.table.columns.get_headers()[i],self.table.cell(web_row[i], Column(headers[i]), Row(i))]))
            else:
                raise ValueError("Value for 'get_row' should string or int")
            return res
        except Exception as ex:
                raise Exception(val + " " + str(ex))

    #def get(self):
    #    return map(lambda self.get_headers())


